File: examples/python/controller.py
#!/usr/bin/env python3
"""
Complete Controller UniFFI Test Suite
Tests all functionality including full workflows with signup, execution, and transfers
"""

import sys
import shutil
import os
from pathlib import Path

# Setup paths
repo_root = Path(__file__).parent.parent.parent
bindings_path = repo_root / "bindings" / "python"
lib_source = repo_root / "target" / "release" / "libcontroller_uniffi.dylib"
lib_dest = bindings_path / "libcontroller_uniffi.dylib"

# Copy library to bindings directory if it doesn't exist or is outdated
if not lib_dest.exists() or lib_source.stat().st_mtime > lib_dest.stat().st_mtime:
    print(f"Copying library from {lib_source} to {lib_dest}...")
    shutil.copy2(lib_source, lib_dest)
    print("✓ Library copied")

# Add the bindings directory to the path
sys.path.insert(0, str(bindings_path))

from controller_uniffi import (
    Owner,
    Controller,
    SessionAccount,
    FieldElement,
    Call,
    SessionPolicy,
    SessionPolicies,
    SignerType,
    Version,
    ControllerError,
    validate_felt,
    get_public_key,
    signer_to_guid,
    get_controller_class_hash,
    controller_has_storage,
)

def generate_stark_private_key():
    """Generate a random Stark private key for testing"""
    import secrets
    return "0x" + secrets.token_hex(32)

def test_utility_functions():
    """Test utility functions"""
    print("\n=== Step 1: Testing Utility Functions ===")
    
    # Test validate_felt
    try:
        result = validate_felt("0x1234")
        print(f"✓ validate_felt: {result}")
    except Exception as e:
        print(f"✗ validate_felt error: {e}")
    
    # Test get_public_key
    try:
        private_key = "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
        public_key = get_public_key(private_key)
        print(f"✓ get_public_key: {public_key[:20]}...")
    except Exception as e:
        print(f"✗ get_public_key error: {e}")
    
    # Test signer_to_guid
    try:
        guid = signer_to_guid(private_key)
        print(f"✓ signer_to_guid: {guid[:20]}...")
    except Exception as e:
        print(f"✗ signer_to_guid error: {e}")
    
    # Test get_controller_class_hash
    try:
        for version in [Version.V1_0_9, Version.LATEST]:
            class_hash = get_controller_class_hash(version)
            print(f"✓ get_controller_class_hash({version}): {class_hash[:20]}...")
    except Exception as e:
        print(f"✗ get_controller_class_hash error: {e}")

def test_owner_creation(private_key):
    """Test Owner creation"""
    print("\n=== Step 2: Creating Owner ===")
    print(f"🔑 Using private key: {private_key}")
    
    try:
        owner = Owner(private_key)
        print("✅ Owner created successfully")
        return owner
    except Exception as e:
        print(f"❌ Owner creation error: {e}")
        return None

def test_controller_creation(owner, private_key):
    """Test Controller creation with multiple constructors"""
    print("\n=== Step 3: Creating Controller ===")
    
    # Configuration
    app_id = "test_app_python"
    username = f"pythonuser-{os.urandom(4).hex()}"
    rpc_url = "https://api.cartridge.gg/x/starknet/sepolia"
    chain_id = "0x534e5f5345504f4c4941"  # SN_SEPOLIA
    
    try:
        # Get the latest controller class hash
        print("📋 Getting controller class hash...")
        class_hash = get_controller_class_hash(Version.LATEST)
        print(f"📄 Class hash: {class_hash}")
        
        # Create headless controller
        print("🎮 Creating headless controller...")
        controller = Controller.new_headless(
            app_id=app_id,
            username=username,
            class_hash=class_hash,
            rpc_url=rpc_url,
            owner=owner,
            chain_id=chain_id
        )
        print("✅ Controller created successfully")
        
        # Get controller information
        print("\n📊 Controller Information:")
        print(f"  📍 Address: {controller.address()}")
        print(f"  👤 Username: {controller.username()}")
        print(f"  🆔 App ID: {controller.app_id()}")
        print(f"  ⛓️  Chain ID: {controller.chain_id()}")
        
        return controller
    except Exception as e:
        print(f"❌ Controller creation error: {e}")
        return None

def test_controller_storage():
    """Test controller storage functions"""
    print("\n=== Step 4: Testing Controller Storage ===")
    
    try:
        # Check if storage exists for non-existent app
        has_storage = controller_has_storage("nonexistent_app")
        print(f"✓ controller_has_storage('nonexistent_app'): {has_storage}")
        
        # Try to load from storage (should fail)
        try:
            stored = Controller.from_storage("nonexistent_app")
            print(f"⚠️  Unexpectedly found stored controller")
        except ControllerError as e:
            print(f"✓ Controller.from_storage correctly throws error for non-existent app")
    except Exception as e:
        print(f"✗ Storage test error: {e}")

def test_controller_signup(controller):
    """Test controller signup"""
    print("\n=== Step 5: Testing Signup ===")
    
    try:
        controller.signup(
            signer_type=SignerType.STARKNET,
            session_expiration=None,
            cartridge_api_url=None
        )
        print("✅ Signup successful")
        return True
    except Exception as e:
        print(f"⚠️  Signup error: {e}")
        print("   (This is expected if the account already exists or is not funded)")
        return False

def test_controller_execution(controller):
    """Test controller transaction execution"""
    print("\n=== Step 6: Testing Transaction Execution ===")
    
    try:
        # Create a test call
        print("📦 Creating test transaction...")
        call = Call(
            contract_address="0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",  # ETH contract
            entrypoint="approve",
            calldata=[
                "0x1234567890abcdef",  # spender
                "0x100",  # amount low
                "0x0"     # amount high
            ]
        )
        
        print("🚀 Executing transaction...")
        tx_hash = controller.execute([call])
        print(f"✅ Transaction executed: {tx_hash}")
        return True
    except Exception as e:
        print(f"❌ Execution error: {e}")
        print("   (This is expected if the account is not deployed or not funded)")
        return False

def test_controller_transfer(controller):
    """Test simplified transfer function"""
    print("\n=== Step 7: Testing Transfer ===")
    
    try:
        recipient = "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
        amount = "0x1"  # 1 wei
        
        print(f"💰 Transferring {amount} to {recipient[:20]}...")
        tx_hash = controller.transfer(recipient, amount)
        print(f"✅ Transfer successful: {tx_hash}")
        return True
    except Exception as e:
        print(f"❌ Transfer error: {e}")
        print("   (This is expected if the account is not deployed or not funded)")
        return False

def test_session_account():
    """Test SessionAccount creation and functionality"""
    print("\n=== Step 8: Testing SessionAccount ===")
    
    try:
        # Create session policies
        print("📜 Creating session policies...")
        policy = SessionPolicy(
            contract_address="0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
            entrypoint="transfer"
        )
        
        policies = SessionPolicies(
            policies=[policy],
            max_fee="0x100000"
        )
        
        # Create session account
        print("🔐 Creating session account...")
        session = SessionAccount(
            rpc_url="https://api.cartridge.gg/x/starknet/sepolia",
            private_key=generate_stark_private_key(),
            address="0x5678",
            owner_guid="0x9abc",
            chain_id="0x534e5f5345504f4c4941",
            policies=policies,
            session_expiration=1735689600
        )
        print("✅ SessionAccount created successfully")
        return session
    except Exception as e:
        print(f"❌ SessionAccount creation error: {e}")
        return None

def test_session_account_from_subscribe():
    """Test SessionAccount.create_from_subscribe"""
    print("\n=== Step 9: Testing SessionAccount.create_from_subscribe ===")
    
    try:
        policy = SessionPolicy(
            contract_address="0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
            entrypoint="transfer"
        )
        
        policies = SessionPolicies(
            policies=[policy],
            max_fee="0x100000"
        )
        
        print("🌐 Attempting to create session from GraphQL API...")
        session = SessionAccount.create_from_subscribe(
            private_key=generate_stark_private_key(),
            policies=policies,
            rpc_url="https://api.cartridge.gg/x/starknet/sepolia",
            cartridge_api_url="https://x.cartridge.gg/graphql"
        )
        print("✅ SessionAccount.create_from_subscribe succeeded")
        return session
    except ControllerError as e:
        print(f"⚠️  SessionAccount.create_from_subscribe failed (expected without real session):")
        print(f"   {str(e)[:100]}...")
        return None

def test_session_execution(session):
    """Test session account execution"""
    if not session:
        print("\n=== Step 10: Skipping Session Execution (no session) ===")
        return
    
    print("\n=== Step 10: Testing Session Execution ===")
    
    try:
        call = Call(
            contract_address="0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
            entrypoint="transfer",
            calldata=[
                "0x1234567890abcdef",
                "0x1",
                "0x0"
            ]
        )
        
        print("🚀 Executing with session...")
        tx_hash = session.execute([call])
        print(f"✅ Session execution successful: {tx_hash}")
    except Exception as e:
        print(f"❌ Session execution error: {e}")

def test_error_handling(controller):
    """Test error handling and error messages"""
    print("\n=== Step 11: Testing Error Handling ===")
    
    try:
        # Try an invalid operation
        controller.switch_chain("invalid_url")
    except Exception as e:
        print(f"✓ Error correctly caught: {type(e).__name__}")
        
        # Get error message
        try:
            error_msg = controller.error_message()
            if error_msg:
                print(f"✓ Error message available: {error_msg[:50]}...")
            
            # Clear error
            controller.clear_last_error()
            print("✓ Error cleared")
        except:
            pass

def main():
    print("=" * 70)
    print("  🚀 COMPLETE CONTROLLER-UNIFFI TEST SUITE")
    print("  Testing all functionality with full workflows")
    print("=" * 70)
    
    # Generate a test private key
    private_key = generate_stark_private_key()
    
    # Run all tests in sequence
    test_utility_functions()
    
    owner = test_owner_creation(private_key)
    if not owner:
        print("\n❌ Cannot continue without owner")
        return 1
    
    controller = test_controller_creation(owner, private_key)
    if not controller:
        print("\n❌ Cannot continue without controller")
        return 1
    
    test_controller_storage()
    test_controller_signup(controller)
    test_controller_execution(controller)
    test_controller_transfer(controller)
    
    session = test_session_account()
    session_from_api = test_session_account_from_subscribe()
    test_session_execution(session_from_api if session_from_api else session)
    
    test_error_handling(controller)
    
    print("\n" + "=" * 70)
    print("  ✅ ALL TESTS COMPLETED")
    print("  All modules and workflows tested!")
    print("=" * 70)
    print()
    print("📝 Note: Some operations may fail if the account is not deployed/funded.")
    print("   This is expected behavior for a test environment.")
    print()
    
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: examples/swift/controller.swift
/**
 * Complete Controller UniFFI Test Suite
 * Tests all functionality including full workflows with signup, execution, and transfers
 */

import Foundation

// ANSI color codes
let GREEN = "\u{001B}[0;32m"
let RED = "\u{001B}[0;31m"
let BLUE = "\u{001B}[0;34m"
let YELLOW = "\u{001B}[1;33m"
let NC = "\u{001B}[0m"

func printSection(_ title: String) {
    print("\n\(BLUE)=== \(title) ===\(NC)")
}

func printSuccess(_ message: String) {
    print("\(GREEN)✓ \(message)\(NC)")
}

func printError(_ message: String) {
    print("\(RED)❌ \(message)\(NC)")
}

func printWarning(_ message: String) {
    print("\(YELLOW)⚠️  \(message)\(NC)")
}

func printInfo(_ message: String) {
    print("  \(message)")
}

// Generate random private key for testing
func generateStarkPrivateKey() -> String {
    var bytes = [UInt8](repeating: 0, count: 32)
    _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
    return "0x" + bytes.map { String(format: "%02x", $0) }.joined()
}

// MARK: - Step 1: Test Utility Functions

func testUtilityFunctions() {
    printSection("Step 1: Testing Utility Functions")
    
    do {
        _ = try validateFelt(felt: "0x1234")
        printSuccess("validate_felt: works")
    } catch {
        printError("validate_felt error: \(error)")
    }
    
    do {
        let privateKey = "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
        let publicKey = try getPublicKey(privateKey: privateKey)
        printSuccess("get_public_key: \(String(publicKey.prefix(20)))...")
    } catch {
        printError("get_public_key error: \(error)")
    }
    
    do {
        let privateKey = "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
        let guid = try signerToGuid(privateKey: privateKey)
        printSuccess("signer_to_guid: \(String(guid.prefix(20)))...")
    } catch {
        printError("signer_to_guid error: \(error)")
    }
    
    do {
        for version in [Version.v109, Version.latest] {
            let classHash = try getControllerClassHash(version: version)
            printSuccess("get_controller_class_hash(\(version)): \(String(classHash.prefix(20)))...")
        }
    } catch {
        printError("get_controller_class_hash error: \(error)")
    }
}

// MARK: - Step 2: Owner Creation

func testOwnerCreation(privateKey: String) -> Owner? {
    printSection("Step 2: Creating Owner")
    print("🔑 Using private key: \(privateKey)")
    
    do {
        let owner = try Owner(privateKey: privateKey)
        printSuccess("Owner created successfully")
        return owner
    } catch {
        printError("Owner creation error: \(error)")
        return nil
    }
}

// MARK: - Step 3: Controller Creation

func testControllerCreation(owner: Owner, privateKey: String) -> Controller? {
    printSection("Step 3: Creating Controller")
    
    let appId = "test_app_swift"
    let username = "swiftuser-\(UUID().uuidString.prefix(8))"
    let rpcUrl = "https://api.cartridge.gg/x/starknet/sepolia"
    let chainId = "0x534e5f5345504f4c4941"
    
    do {
        print("📋 Getting controller class hash...")
        let classHash = try getControllerClassHash(version: .latest)
        print("📄 Class hash: \(classHash)")
        
        print("🎮 Creating headless controller...")
        let controller = try Controller.newHeadless(
            appId: appId,
            username: username,
            classHash: classHash,
            rpcUrl: rpcUrl,
            owner: owner,
            chainId: chainId
        )
        printSuccess("Controller created successfully")
        
        print("\n📊 Controller Information:")
        printInfo("📍 Address: \(try controller.address())")
        printInfo("👤 Username: \(try controller.username())")
        printInfo("🆔 App ID: \(try controller.appId())")
        printInfo("⛓️  Chain ID: \(try controller.chainId())")
        
        return controller
    } catch {
        printError("Controller creation error: \(error)")
        return nil
    }
}

// MARK: - Step 4: Controller Storage

func testControllerStorage() {
    printSection("Step 4: Testing Controller Storage")
    
    do {
        let hasStorage = try controllerHasStorage(appId: "nonexistent_app")
        printSuccess("controller_has_storage('nonexistent_app'): \(hasStorage)")
        
        do {
            _ = try Controller.fromStorage(appId: "nonexistent_app")
            printWarning("Unexpectedly found stored controller")
        } catch {
            printSuccess("Controller.fromStorage correctly throws error for non-existent app")
        }
    } catch {
        printError("Storage test error: \(error)")
    }
}

// MARK: - Step 5: Signup

func testControllerSignup(controller: Controller) -> Bool {
    printSection("Step 5: Testing Signup")
    
    do {
        try controller.signup(
            signerType: .starknet,
            sessionExpiration: nil,
            cartridgeApiUrl: nil
        )
        printSuccess("Signup successful")
        return true
    } catch {
        printWarning("Signup error: \(error)")
        printInfo("(This is expected if the account already exists or is not funded)")
        return false
    }
}

// MARK: - Step 6: Transaction Execution

func testControllerExecution(controller: Controller) -> Bool {
    printSection("Step 6: Testing Transaction Execution")
    
    do {
        print("📦 Creating test transaction...")
        let call = Call(
            contractAddress: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
            entrypoint: "approve",
            calldata: [
                "0x1234567890abcdef",
                "0x100",
                "0x0"
            ]
        )
        
        print("🚀 Executing transaction...")
        let txHash = try controller.execute(calls: [call])
        printSuccess("Transaction executed: \(txHash)")
        return true
    } catch {
        printError("Execution error: \(error)")
        printInfo("(This is expected if the account is not deployed or not funded)")
        return false
    }
}

// MARK: - Step 7: Transfer

func testControllerTransfer(controller: Controller) -> Bool {
    printSection("Step 7: Testing Transfer")
    
    do {
        let recipient = "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef"
        let amount = "0x1"
        
        print("💰 Transferring \(amount) to \(String(recipient.prefix(20)))...")
        let txHash = try controller.transfer(recipient: recipient, amount: amount)
        printSuccess("Transfer successful: \(txHash)")
        return true
    } catch {
        printError("Transfer error: \(error)")
        printInfo("(This is expected if the account is not deployed or not funded)")
        return false
    }
}

// MARK: - Step 8: SessionAccount

func testSessionAccount() -> SessionAccount? {
    printSection("Step 8: Testing SessionAccount")
    
    do {
        print("📜 Creating session policies...")
        let policy = SessionPolicy(
            contractAddress: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
            entrypoint: "transfer"
        )
        
        let policies = SessionPolicies(
            policies: [policy],
            maxFee: "0x100000"
        )
        
        print("🔐 Creating session account...")
        let session = try SessionAccount(
            rpcUrl: "https://api.cartridge.gg/x/starknet/sepolia",
            privateKey: generateStarkPrivateKey(),
            address: "0x5678",
            ownerGuid: "0x9abc",
            chainId: "0x534e5f5345504f4c4941",
            policies: policies,
            sessionExpiration: 1735689600
        )
        printSuccess("SessionAccount created successfully")
        return session
    } catch {
        printError("SessionAccount creation error: \(error)")
        return nil
    }
}

// MARK: - Step 9: SessionAccount.createFromSubscribe

func testSessionAccountFromSubscribe() -> SessionAccount? {
    printSection("Step 9: Testing SessionAccount.createFromSubscribe")
    
    do {
        let policy = SessionPolicy(
            contractAddress: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
            entrypoint: "transfer"
        )
        
        let policies = SessionPolicies(
            policies: [policy],
            maxFee: "0x100000"
        )
        
        print("🌐 Attempting to create session from GraphQL API...")
        let session = try SessionAccount.createFromSubscribe(
            privateKey: generateStarkPrivateKey(),
            policies: policies,
            rpcUrl: "https://api.cartridge.gg/x/starknet/sepolia",
            cartridgeApiUrl: "https://x.cartridge.gg/graphql"
        )
        printSuccess("SessionAccount.createFromSubscribe succeeded")
        return session
    } catch {
        printWarning("SessionAccount.createFromSubscribe failed (expected without real session):")
        printInfo(String(describing: error).prefix(100) + "...")
        return nil
    }
}

// MARK: - Step 10: Session Execution

func testSessionExecution(session: SessionAccount?) {
    if session == nil {
        printSection("Step 10: Skipping Session Execution (no session)")
        return
    }
    
    printSection("Step 10: Testing Session Execution")
    
    do {
        let call = Call(
            contractAddress: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
            entrypoint: "transfer",
            calldata: [
                "0x1234567890abcdef",
                "0x1",
                "0x0"
            ]
        )
        
        print("🚀 Executing with session...")
        let txHash = try session!.execute(calls: [call])
        printSuccess("Session execution successful: \(txHash)")
    } catch {
        printError("Session execution error: \(error)")
    }
}

// MARK: - Step 11: Error Handling

func testErrorHandling(controller: Controller) {
    printSection("Step 11: Testing Error Handling")
    
    do {
        try controller.switchChain(rpcUrl: "invalid_url")
    } catch {
        printSuccess("Error correctly caught: \(type(of: error))")
        
        do {
            let errorMsg = try controller.errorMessage()
            if !errorMsg.isEmpty {
                printSuccess("Error message available: \(String(errorMsg.prefix(50)))...")
            }
            
            controller.clearLastError()
            printSuccess("Error cleared")
        } catch {
            // Ignore
        }
    }
}

// MARK: - Main

func runTests() {
    print(String(repeating: "=", count: 70))
    print("  🚀 COMPLETE CONTROLLER-UNIFFI SWIFT TEST SUITE")
    print("  Testing all functionality with full workflows")
    print(String(repeating: "=", count: 70))
    
    let privateKey = generateStarkPrivateKey()
    
    testUtilityFunctions()
    
    guard let owner = testOwnerCreation(privateKey: privateKey) else {
        printError("Cannot continue without owner")
        exit(1)
    }
    
    guard let controller = testControllerCreation(owner: owner, privateKey: privateKey) else {
        printError("Cannot continue without controller")
        exit(1)
    }
    
    testControllerStorage()
    _ = testControllerSignup(controller: controller)
    _ = testControllerExecution(controller: controller)
    _ = testControllerTransfer(controller: controller)
    
    let session = testSessionAccount()
    let sessionFromApi = testSessionAccountFromSubscribe()
    testSessionExecution(session: sessionFromApi ?? session)
    
    testErrorHandling(controller: controller)
    
    print("\n" + String(repeating: "=", count: 70))
    print("  \(GREEN)✅ ALL TESTS COMPLETED\(NC)")
    print("  All modules and workflows tested!")
    print(String(repeating: "=", count: 70))
    print()
    print("📝 Note: Some operations may fail if the account is not deployed/funded.")
    print("   This is expected behavior for a test environment.")
    print()
}

// Run the tests
runTests()
exit(0)
